### PR TITLE
[DOCS] Semantics of negation.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/3860

### CODE REVIEW

File: docs/types.rst [hunk 1]
```diff
@@ -82,10 +82,22 @@ Addition, Subtraction and Multiplication
...
 into account when designing safe smart contracts.
 
+The expression ``-x`` is equivalent to ``(T(0) - x)`` where
+``T`` is the type of ``x``. This means that ``-x`` will not be negative
+if the type of ``x`` is an unsigned integer type. Also, ``-x`` can be
+positive if ``x`` is negative. There is another caveat also resulting
+from two's complement representation::
+
+    int x = -2**255;
+    assert(-x == -x);
```
Comment (axic) [inline docs/types.rst]: What exactly is this testing?

Comment (erak) [reply docs/types.rst]: I think it should be ``assert(-x == x)``. @chriseth ?

**Comment by axic** [comment]
Has a conflict now, needs a rebase :(

**Comment by chriseth** [comment]
Updated.

File: docs/types.rst [hunk 2]
```diff
@@ -82,10 +82,23 @@ Addition, Subtraction and Multiplication
...
 into account when designing safe smart contracts.
 
+The expression ``-x`` is equivalent to ``(T(0) - x)`` where
+``T`` is the type of ``x``. This means that ``-x`` will not be negative
+if the type of ``x`` is an unsigned integer type. Also, ``-x`` can be
+positive if ``x`` is negative. There is another caveat also resulting
+from two's complement representation::
+
+    int x = -2**255;
+    assert(-x == x);
```
Comment (axic) [inline docs/types.rst:96]: Hm, just realised, do we have a test case for this in the EndToEnd suite?

**Review by axic — APPROVED** [review]

File: docs/types.rst (hunk 2 above)
Comment (chriseth) [reply docs/types.rst:96]: We could add it, but that's just EVM's semantics.

Comment (axic) [reply docs/types.rst:96]: Yeah but we're perhaps going to have runtime overflow checks, etc., so better to add a test case when we see it.

**Pull request merged**